File: thinking-categorically/cartesian-closed-category/hom-set/hom-set.ts
// The Hom-set Hom(A, B) corresponds to function type (A) => B
type Hom<A, B> = (a: A) => B;

// In a CCC: Hom(A, B) ≅ B^A (exponential object)
type Exponential<A, B> = Hom<A, B>;  // They're the same!

// The currying isomorphism: Hom(A × B, C) ≅ Hom(A, Hom(B, C))
type Product<A, B> = [A, B];
type CurryingIsomorphism<A, B, C> = {
  left: Hom<Product<A, B>, C>;      // Functions from pairs
  right: Hom<A, Hom<B, C>>;         // Curried functions
};

// Demonstrating the isomorphism
const curryingDemo = () => {
  // Left side: Hom([number, number], number)
  const multiply: Hom<Product<number, number>, number> =
    ([x, y]) => x * y;

  // Right side: Hom(number, Hom(number, number))
  const curryMultiply: Hom<number, Hom<number, number>> =
    (x) => (y) => x * y;

  // The isomorphism in action
  console.log(multiply([6, 7]));      // 42
  console.log(curryMultiply(6)(7));   // 42

  // curry/uncurry witness the isomorphism
  const curry = <A, B, C>(f: Hom<Product<A, B>, C>): Hom<A, Hom<B, C>> =>
    (a) => (b) => f([a, b]);

  const uncurry = <A, B, C>(f: Hom<A, Hom<B, C>>): Hom<Product<A, B>, C> =>
    ([a, b]) => f(a)(b);

  const curried = curry(multiply);
  const uncurried = uncurry(curryMultiply);

  console.log(curried(6)(7));        // 42
  console.log(uncurried([6, 7]));    // 42
};

// Higher-order functions manipulate Hom-sets
const mapHom = <A, B, C>(
  f: Hom<B, C>,
  functions: Hom<A, B>[]
): Hom<A, C>[] =>
  functions.map(g => (x: A) => f(g(x)));  // Function composition

// Example: Working with collections of functions
const functionManipulation = () => {
  // A collection of functions (elements of Hom(number, number))
  const numberFunctions: Hom<number, number>[] = [
    x => x + 1,
    x => x * 2,
    x => x * x
  ];

  // Transform them with another function (Hom(number, string))
  const toString: Hom<number, string> = x => `Result: ${x}`;

  // Get new functions (elements of Hom(number, string))
  const stringFunctions = mapHom(toString, numberFunctions);

  // Test the transformed functions
  stringFunctions.forEach((f, i) => {
    console.log(`Function ${i}(5) = ${f(5)}`);
    // "Function 0(5) = Result: 6"
    // "Function 1(5) = Result: 10"
    // "Function 2(5) = Result: 25"
  });
};

// Exponential objects as function spaces
const exponentialObjects = () => {
  // String^number represents all functions from number to string
  const stringFromNumber: Hom<number, string>[] = [
    x => x.toString(),
    x => `The number ${x}`,
    x => "*".repeat(x),
    x => x < 0 ? "negative" : x > 0 ? "positive" : "zero"
  ];

  // Functions are first-class - can be stored, passed, returned
  const selectFunction = (index: number): Hom<number, string> =>
    stringFromNumber[index] || (x => `Unknown: ${x}`);

  const selectedFunc = selectFunction(2);
  console.log(selectedFunc(5)); // "*****"

  // Function composition creates new exponential objects
  const addPrefix: Hom<string, string> = s => `Output: ${s}`;
  const composedFunctions = stringFromNumber.map(f =>
    (x: number) => addPrefix(f(x))
  );

  console.log(composedFunctions[1](42)); // "Output: The number 42"
};


// Demonstrating the category structure of Hom-sets
const homCategory = () => {
  // Objects: Types (A, B, C, ...)
  // Morphisms: Natural transformations between Hom-sets

  // Identity natural transformation: Hom(A,B) -> Hom(A,B)
  const idHom = <A, B>(f: Hom<A, B>): Hom<A, B> => f;

  // Example natural transformation: pre-composition
  // Given g: A -> B, get natural transformation Hom(B,C) -> Hom(A,C)
  const preComposition = <A, B, C>(g: Hom<A, B>) =>
    (f: Hom<B, C>): Hom<A, C> => (x: A) => f(g(x));

  // Example usage
  const double: Hom<number, number> = x => x * 2;
  const toString: Hom<number, string> = x => `Value: ${x}`;
  const preComposeWithDouble = preComposition(double);
  const composedFunction = preComposeWithDouble(toString);
  console.log(`Pre-composed function(5) = ${composedFunction(5)}`); // "Value: 10"

  console.log("Hom-sets form a category with natural transformations");
};

// The fundamental insight: Hom(A,B) ≅ B^A
const fundamentalTheorem = () => {
  console.log("=== Fundamental Theorem: Hom(A,B) ≅ B^A ===");
  console.log("In TypeScript (modeling a Cartesian Closed Category):");
  console.log("- Types are objects");
  console.log("- Function types (A) => B are morphisms in Hom(A,B)");
  console.log("- Function types (A) => B are also objects (exponentials B^A)");
  console.log("- This duality enables functional programming patterns\n");

  // Demonstrate that function types are both morphisms AND objects
  const functionAsMorphism: Hom<number, string> = x => x.toString();  // Morphism in Hom(number, string)

  // The same function type as objects that can be manipulated
  const functionsAsObjects: Hom<number, string>[] = [
    functionAsMorphism,  // Our function is now data!
    x => `Number: ${x}`,
    x => "*".repeat(Math.max(0, x)),
    x => x < 0 ? "negative" : x > 0 ? "positive" : "zero"
  ];

  console.log("Functions as objects - stored in an array:");
  functionsAsObjects.forEach((func, index) => {
    console.log(`  Function ${index}(3) = "${func(3)}"`);
  });

  // Higher-order function that takes functions as arguments (objects)
  // and returns a function (object) - this is only possible because
  // Hom(A,B) ≅ B^A makes functions first-class
  const testFunction: Hom<Hom<number, string>, string> = f => f(42);

  console.log("\nHigher-order function applying each function to 42:");
  functionsAsObjects.forEach((func, index) => {
    console.log(`  Testing function ${index}: "${testFunction(func)}"`);
  });

  // Functions can be returned from other functions (function factories)
  const createFormatter = (prefix: string): Hom<number, string> =>
    x => `${prefix}: ${x}`;

  const errorFormatter = createFormatter("ERROR");
  const infoFormatter = createFormatter("INFO");

  console.log("\nFunction factories creating new functions:");
  console.log(`  ${errorFormatter(404)}`);  // "ERROR: 404"
  console.log(`  ${infoFormatter(200)}`);   // "INFO: 200"

  // The key insight: functions are simultaneously structure and data
  console.log("\n🔑 Key Insight:");
  console.log("Functions are simultaneously:");
  console.log("- MORPHISMS: They represent computational relationships");
  console.log("- OBJECTS: They can be manipulated as first-class data");
  console.log("This dual nature is what makes functional programming powerful!");
};

curryingDemo();
functionManipulation();
exponentialObjects();
homCategory();
fundamentalTheorem();

File: thinking-categorically/cartesian-closed-category/hom-set/hom-set.cs
using System;
using System.Collections.Generic;
using System.Linq;

// The Hom-set Hom(A, B) corresponds to delegate types
public delegate TResult Hom<T, TResult>(T input);

// In a CCC: Hom(A, B) ≅ B^A (exponential object)
public delegate TResult Exponential<T, TResult>(T input);

// Product type for demonstrating currying
public struct Product<A, B>
{
    public A First { get; }
    public B Second { get; }

    public Product(A first, B second)
    {
        First = first;
        Second = second;
    }
}

public static class HomSetExamples
{
    // The currying isomorphism: Hom(A × B, C) ≅ Hom(A, Hom(B, C))
    public static void CurryingIsomorphism()
    {
        Console.WriteLine("=== Currying Isomorphism ===");

        // Left side: Hom(Product<int, int>, int)
        Hom<Product<int, int>, int> multiply =
            pair => pair.First * pair.Second;

        // Right side: Hom(int, Hom(int, int))
        Hom<int, Hom<int, int>> curryMultiply =
            x => y => x * y;

        // Demonstrate the isomorphism
        var testPair = new Product<int, int>(6, 7);
        Console.WriteLine($"multiply(6,7) = {multiply(testPair)}");           // 42
        Console.WriteLine($"curryMultiply(6)(7) = {curryMultiply(6)(7)}");   // 42

        // curry/uncurry witness the isomorphism
        var curried = Curry(multiply);
        var uncurried = Uncurry(curryMultiply);

        Console.WriteLine($"curried(6)(7) = {curried(6)(7)}");               // 42
        Console.WriteLine($"uncurried(6,7) = {uncurried(testPair)}");        // 42
    }

    // Curry: Hom(A × B, C) -> Hom(A, Hom(B, C))
    public static Hom<A, Hom<B, C>> Curry<A, B, C>(Hom<Product<A, B>, C> f) =>
        a => b => f(new Product<A, B>(a, b));

    // Uncurry: Hom(A, Hom(B, C)) -> Hom(A × B, C)
    public static Hom<Product<A, B>, C> Uncurry<A, B, C>(Hom<A, Hom<B, C>> f) =>
        pair => f(pair.First)(pair.Second);

    // Higher-order functions manipulate Hom-sets
    public static IEnumerable<Hom<A, C>> MapHom<A, B, C>(
        Hom<B, C> f,
        IEnumerable<Hom<A, B>> functions) =>
        functions.Select(g => new Hom<A, C>(x => f(g(x))));

    // Example: Working with collections of functions
    public static void FunctionManipulation()
    {
        Console.WriteLine("\n=== Function Manipulation ===");

        // Collection of functions (elements of Hom(int, int))
        var numberFunctions = new List<Hom<int, int>>
        {
            x => x + 1,
            x => x * 2,
            x => x * x
        };

        // Transform with another function (Hom(int, string))
        Hom<int, string> toString = x => $"Result: {x}";

        // Get new functions (elements of Hom(int, string))
        var stringFunctions = MapHom(toString, numberFunctions).ToList();

        // Test the transformed functions
        for (int i = 0; i < stringFunctions.Count; i++)
        {
            Console.WriteLine($"Function {i}(5) = {stringFunctions[i](5)}");
        }
    }

    // Exponential objects as function spaces
    public static void ExponentialObjects()
    {
        Console.WriteLine("\n=== Exponential Objects ===");

        // string^int represents all functions from int to string
        var stringFromInt = new List<Exponential<int, string>>
        {
            x => x.ToString(),
            x => $"The number {x}",
            x => new string('*', Math.Max(0, x)),
            x => x < 0 ? "negative" : x > 0 ? "positive" : "zero"
        };

        // Functions are first-class - can be stored, passed, returned
        Hom<int, Exponential<int, string>> selectFunction =
            index => index < stringFromInt.Count ? stringFromInt[index] :
                     (x => $"Unknown: {x}");

        var selectedFunc = selectFunction(2);
        Console.WriteLine($"Selected function(5) = {selectedFunc(5)}"); // *****

        // Function composition creates new exponential objects
        Hom<string, string> addPrefix = s => $"Output: {s}";
        var composedFunctions = stringFromInt
            .Select(f => new Exponential<int, string>(x => addPrefix(f(x))))
            .ToList();

        Console.WriteLine($"Composed function(3) = {composedFunctions[1](3)}");
        // "Output: The number 3"
    }

    // Natural transformations between Hom-sets
    // eta: ∀A,B. A -> Hom(B, A)  (constant function)
    public static Hom<B, A> Eta<A, B>(A value) => _ => value;

    // Demonstrating the category structure of Hom-sets
    public static void HomCategory()
    {
        Console.WriteLine("\n=== Hom-sets Category ===");

        // Identity natural transformation: Hom(A,B) -> Hom(A,B)
        Hom<Hom<int, string>, Hom<int, string>> idHom = f => f;

        // Example of natural transformation: pre-composition
        // Given g: A -> B, get (g*): Hom(B,C) -> Hom(A,C)
        // where (g*)(f) = f ∘ g
        Hom<int, int> doubleIt = x => x * 2;
        Hom<Hom<int, string>, Hom<int, string>> preCompose =
            f => x => f(doubleIt(x));

        Hom<int, string> originalFunction = x => $"Value: {x}";
        var transformedFunction = preCompose(originalFunction);

        Console.WriteLine($"Original(5) = {originalFunction(5)}");      // "Value: 5"
        Console.WriteLine($"Transformed(5) = {transformedFunction(5)}"); // "Value: 10"

        Console.WriteLine("Hom-sets form a category with natural transformations as morphisms");
    }

    // The fundamental insight: Hom(A,B) ≅ B^A
    public static void FundamentalTheorem()
    {
        // Demonstrate that function types are both morphisms AND objects
        Hom<int, string> functionAsMorphism = x => x.ToString();  // Morphism in Hom(int, string)

        // The same function type as an object that can be manipulated
        var functionsAsObjects = new List<Hom<int, string>>
        {
            functionAsMorphism,  // Our function is now data!
            x => $"Number: {x}",
            x => new string('X', x)
        };

        // Higher-order function that takes functions as arguments (objects)
        // and returns a function (object) - this is only possible because
        // Hom(A,B) ≅ B^A makes functions first-class
        Hom<Hom<int, string>, string> testFunction = f => f(42);

        foreach (var func in functionsAsObjects)
        {
            Console.WriteLine($"Testing function: {testFunction(func)}");
        }
    }
}

class Program
{
    static void Main()
    {
        Console.WriteLine("Hom-sets Connection to Cartesian Closed Categories\n");

        HomSetExamples.CurryingIsomorphism();
        HomSetExamples.FunctionManipulation();
        HomSetExamples.ExponentialObjects();
        HomSetExamples.HomCategory();
        HomSetExamples.FundamentalTheorem();
    }
}